src/game/spawn.rs

```diff
@@ -0,0 +1,835 @@
+					*session,
+					VertexPosition(id),
+					TransformBundle::default(),
+					VisibilityBundle::default(),
+				))
+				.id(),
+		});
+		let glyph_id = data.glyph.map(|glyph| match glyph {
+			GlyphData::Flag => commands
+				.spawn((
+					Glyph,
+					Goal,
+					ThingData::Glyph(glyph),
+					*session,
+					VertexPosition(id),
+					TransformBundle::default(),
+					VisibilityBundle::default(),
+				))
+				.id(),
+			GlyphData::Button(None) => commands
+				.spawn((
+					Glyph,
+					BoxSlot,
+					ThingData::Glyph(glyph),
+					*session,
+					VertexPosition(id),
+					TransformBundle::default(),
+					VisibilityBundle::default(),
+				))
+				.id(),
+			GlyphData::Button(Some(color_data)) => commands
+				.spawn((
+					Glyph,
+					BoxSlot,
+					color_data,
+					ThingData::Glyph(glyph),
+					*session,
+					VertexPosition(id),
+					TransformBundle::default(),
+					VisibilityBundle::default(),
+				))
+				.id(),
+		});
+		commands
+			.entity(id)
+			.insert((PlacedObject(object_id), PlacedGlyph(glyph_id)));
+	}
+}
+
+fn set_vertex_transforms(mut query: Query<(&VertexData, &mut Transform), Added<VertexData>>) {
+	for (data, mut transform) in &mut query {
+		transform.translation.x = data.position.x;
+		transform.translation.y = data.position.y;
+	}
+}
+
+fn set_thing_transforms(
+	mut things_q: Query<(&VertexPosition, &mut Transform), Added<VertexPosition>>,
+	vertices_q: Query<&VertexData>,
+) {
+	for (vertex, mut transform) in &mut things_q {
+		let vertex_data = vertices_q
+			.get(vertex.0)
+			.expect("Parent of ThingData does not have VertexData component");
+		transform.translation.x = vertex_data.position.x;
+		transform.translation.y = vertex_data.position.y;
+	}
+}
+
+fn set_cycle_transforms(
+	mut query: Query<(&CyclePlacement, &mut Transform), Added<CyclePlacement>>,
+) {
+	for (placement, mut transform) in &mut query {
+		transform.translation.x = placement.position.x;
+		transform.translation.y = placement.position.y;
+	}
+}
+
+fn create_vertex_visuals(
+	mut commands: Commands,
+	query: Query<Entity, Added<VertexData>>,
+	materials: Res<GameObjectMaterials>,
+	meshes: Res<GameObjectMeshes>,
+) {
+	for id in &query {
+		commands.entity(id).with_children(|children| {
+			children.spawn(ColorMesh2dBundle {
+				mesh: bevy::sprite::Mesh2dHandle(meshes.vertices.clone_weak()),
+				material: materials.cycle_rings.clone_weak(),
+				transform: Transform::from_translation(Vec3::Z * layers::CYCLE_NODES),
+				..default()
+			});
+		});
+	}
+}
+
+fn create_cycle_visuals(
+	mut commands: Commands,
+	query: Query<(Entity, &CyclePlacement, &CycleTurnability), Added<CyclePlacement>>,
+	palette: Res<ThingPalette>,
+	materials: Res<GameObjectMaterials>,
+	images: Res<HandleMap<ImageKey>>,
+	mut meshes: ResMut<Assets<Mesh>>,
+) {
+	for (id, placement, turnability) in &query {
+		let mesh = Annulus::new(
+			placement.radius - RING_HALF_WIDTH,
+			placement.radius + RING_HALF_WIDTH,
+		)
+		.mesh()
+		.resolution(64)
+		.build();
```

Comment:
We're creating a bunch of possibly duplicate meshes here...
although it might not matter because of Handle<> ref-counted "garbage collection".

Reply:
I think de-duplicating these would be more trouble than it is worth. A repository of assigned meshes (`HashMap<f32, Handle<Mesh>>`) would need to be shared between runs of the event somehow, and the most cycles of equal radius per level is like ten (though, a per-run repository might be enough (all new cycles will be registered by one run of the system) and would be easy enough to implement)

Reply:
Oh right. `f32` is not hashable.